hal/network/ncp_client/sara/sara_ncp_client.cpp

```diff
@@ -1056,33 +1056,40 @@ int SaraNcpClient::selectSimCard(ModemState& state) {
 
     if (ncpId() == PLATFORM_NCP_SARA_R410) {
         int resetCount = 0;
+        const int imsiMaxRetryCount = 3;
+        const int CcidMaxRetryCount = 2;
         // Note: Not failing on AT error on ICCID/IMSI lookup since SIMs have shown strange edge cases
         // where they error for no reason, and hard resetting the modem or power cycling would not clear it.
         //
         // Check if we are using a Twilio Super SIM based on ICCID
         char buf[32] = {};
-        int ccidCount = 0;
+        int ccidCount = 1;
         do {
             memset(buf, 0, sizeof(buf));
             auto lenCcid = getIccidImpl(buf, sizeof(buf));
             if (lenCcid > 5) {
                 netConf_ = networkConfigForIccid(buf, lenCcid);
                 break;
             }
-        } while (++ccidCount < 2);
+        } while (++ccidCount <= CcidMaxRetryCount);
         // If failed above i.e., netConf_ is still not valid, look for network settings based on IMSI
         if (!netConf_.isValid()) {
-            int imsiCount = 0;
+            int imsiCount = 1;
             do {
                 memset(buf, 0, sizeof(buf));
                 auto resp = parser_.sendCommand(UBLOX_CIMI_TIMEOUT, "AT+CIMI");
-                CHECK_PARSER(resp.readLine(buf, sizeof(buf)));
-                const int r = CHECK_PARSER(resp.readResult());
-                if (r == AtResponse::OK) {
-                    netConf_ = networkConfigForImsi(buf, strlen(buf));
-                    break;
+                const int res = resp.readLine(buf, sizeof(buf));
+                // Prevent exiting if there is AT response collected in readLine is an error
+                // After max attempts, error out anyway
+                if (res >= 0 || imsiCount >= imsiMaxRetryCount) {
+                    const int r = CHECK_PARSER(resp.readResult());
```

Comment:
I don't think this call should be conditional. We should always readout final result code, otherwise the parser is left in an invalid state. `CHECK_PARSER` also doesn't care about final result codes and only watches for non-AT errors.

What we should probably add though is a call to `hasNextLine()` before we perform `readLine()` in order not to get an error in the case where we don't get an intermediate response and immediately get final response.

Reply:
Added. Please review. Tested on my Boron. 
```
0000156816 [ncp.at] TRACE: > AT+CIMI
0000156824 [ncp.at] TRACE: < +CME ERROR: SIM wrong
0000156825 [ncp.at] TRACE: > AT+CIMI
0000156827 [ncp.at] TRACE: < +CME ERROR: SIM wrong
0000157084 [ncp.at] TRACE: > AT+CIMI
0000157087 [ncp.at] TRACE: < 001010123456789
0000157087 [ncp.at] TRACE: < OK
```

Reply:
And if it fails init (just hard coded retries to 1 retry only for testing purposes) - it fails this way and we would try the whole thing again.
```
0001700252 [ncp.at] TRACE: > AT+CCID
0001700254 [ncp.at] TRACE: < +CCID: 89860000502000180722
0001700257 [ncp.at] TRACE: < OK
0001700263 [ncp.at] TRACE: > AT+CIMI
0001700265 [ncp.at] TRACE: < +CME ERROR: SIM wrong
0001700268 [ncp.client] ERROR: Failed to perform early initialization
0001700274 [ncp.client] TRACE: Setting UART voltage translator state 0
0001700279 [ncp.client] TRACE: Hard resetting the modem
0001711286 [net.pppncp] TRACE: NCP event 3
0001711286 [net.pppncp] TRACE: NCP power state changed: IF_POWER_STATE_DOWN
```